### DISCUSSION TITLE
Roadmap discussions

### DISCUSSION THREAD

**Discussion by JonPSmith** [JonPSmith/AuthPermissions.AspNetCore#2] [Ideas]
| Updated  | Version| What changed |
| ------------- | ------------- | ------------- |
| 23 Dec 2025 | Version 10.0.0 | Supports .NET 10 |
| 30 Nov 2024 | Version 9.0.0 | Supports .NET 9. Has a BREAKING CHANGE |
| 28 May 2024 | Version 8.0.0 | Cleaner and easier to update to next .NET |
| 09 Feb 2024 | Version 6.2.1 | Minor bug fix |
| 10 Jan 2024 | Version 6.2.0 | Update "Sign up new tenant" feature
| 28 Nov 2023 | Version 6.1.0 | Supports .NET 8 |
| 15 Aug 2023 | Version 6.0.0 | Better sharding support |
| 31 May 2023 | Version 5.0.0 | Supports custom database types |
| 3 Jan 2023 | Version 4.1.0 | Supports multiple languages |
| 21 Nov 2022 | Version 4.0.0 | Supports .NET 7 |
| 18 Oct 2022 | Version 3.5.0 | Minor changes |
| 20 Sept 2022 | Version 3.4.0 | New ["Down for maintenance" feature](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/%22Down-for-maintenance%22-feature)| 
| 16 June 2022 | Version 3.3.0 | [“invite a user”](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Invite-new-user-service) and [“sign up / versioning”](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sign-up-for-a-new-tenant%2C-with-versioning) features |
| 26 April 2022 | Version 3.2.0 | 
| 11 April 2022 | Version 3.1.0 |
| 4 April 2022 | Version 3.0.0 | Added sharding |
| 10 Mar 2022 | Version 2.3.1 |
| 6 Feb 2022 | Version 3 | more bugs / features |
| 19 Jan 2022 | Version 3 | Ideas |
| 3 Jan 2022 | Version 2.1.0 |
| 21 Dec 2021 | Rersion 2.0.0 |
|  4 Dec 2021  | Updated version |   
|  10 Aug 2021 | Initial roadmap |

The AuthPermissions.AspNetCore library (shortened to _AuthP_) started with version 1.0.0, but many features have been added. This discussion is to get feedback on this library and what new features people would like.

_REMEMBER: This is an open-source library which I work on in my "spare time", so don't expect me to produce new features quickly._

## Version 10.0.0

- Updated to .NET 10
- And my [JonPSmith.MultiProjPack}(https://www.nuget.org/packages/JonPSmith.MultiProjPack) had to updated to .NET 10 too. NOTE the old `JonPSmith.MultiProjPack` used .NET 6, which is not supported now.

## Version 9.0.0

Updates AuthP to .NET 9.

BREAKING CHANGE: EF Core 9 has added code to ensure that migrations are executed correctly - see EF Core 9's [updated migrations](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-9.0/whatsnew#migrations). If you have build an multi-tenant application then you need to read the [UpdateToVersion9](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion9.md) on how to update your app.

## Version 8.0.0

Simplified to one NET type, in this case .NET 8. The changed makes it easier to update for future NET releases. See the ["How to update a NuGet library once the author isn’t available"](https://www.thereformedprogrammer.net/how-to-update-a-nuget-library-once-the-author-isnt-available/) article for more information.

## Version 6.2.1

Updated Net.RunMethodsSequentially to fix .NET 8 Postgres bug in the [DistributedLock](https://github.com/madelson/DistributedLock). See [this section](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/SetupAspNetCoreAndDatabase#optional-buildingrunning-your-own-migrate--seeding-code-on-startup) in the documentation about this feature.

## Version 6.2.0 

A small change to the [Sign up for a new tenant, with versioning](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sign-up-for-a-new-tenant%2C-with-versioning) feature to handle fatal code problems, e.g. `Exception`s.

This was because @akema-trebla had problems with errors (see issue #97). Once I got into I decided to redesign the code in the `SignInAndCreateTenant` service.

## Version 6.1.0 

- New version to support NET 8 - now supports NET 6, 7, and 8

## Version 6.0.0

- Significant changes to code to support sharding multi-tenant applications
- BREAKING CHANGES: Replaced the old services for handling sharding entries due to a limitation - see [UpdateToVersion6.md](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion6.md) on how to update your app.
- Improvement: The ability to define a sharding-only mode - see [Configuring sharding > sharding-only mode](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Configuring-sharding#4-select-which-mode-you-want-to-use)
- Fix: The Sign up for a new tenant, with versioning now works with sharding-only tenants.
- Improvement: New service to simplify the create / delete of sharding-only tenants - see [Extra help for shading‐only apps](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Configuring-sharding#4-select-which-mode-you-want-to-use) for how that helps.

See new [Example7 app](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example7.MvcWebApp.ShardingOnly) that provides an example of a multi-tenant application where every tenant has its own database. See the [Sharding explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sharding-explained) document about the new _modes_.

## Version 5.0.0

- BREAKING CHANGE in AuthP’s [sharding / hybrid multi-tenant feature](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#2-defining-the-three-multi-tenant-database-arrangements): If you are using AuthP’s sharding /hybrid features, then look at the [UpdateToVersion5.md file](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion5.md) for what you need to do.
- New feature: You can now use range of database providers (see list later) to use with the AuthP library. The table below shown the database providers AuthP 5.0.0 supports:

| Supported database providers in V5.0.0   | Comment            |
| ---------------------------------------- | ------------------ |
| Microsoft.EntityFrameworkCore.SqlServer  | Built-in           |
| Npgsql.EntityFrameworkCore.PostgreSQL    | Built-in			|
| Microsoft.EntityFrameworkCore.Sqlite	   | See examples in [AuthP.CustomDatabaseExamples](https://github.com/JonPSmith/AuthP.CustomDatabaseExamples) repo |
| Microsoft.EntityFrameworkCore.Cosmos 	   | 					|
| Pomelo.EntityFrameworkCore.MySql 		   | [Pomelo Foundation Project](https://github.com/PomeloFoundation) |
| MySql.EntityFrameworkCore				   | [MySQL project](https://dev.mysql.com/)					|
| Oracle.EntityFrameworkCore			   | [Oracle](https://www.oracle.com/database/technologies/appdev/dotnet.html) |

- This release makes it easier to use, and change, the [sharding / hybrid multi-tenant feature](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#2-defining-the-three-multi-tenant-database-arrangements). See the two items below:
   - _Easier to use_: there is a new extension method called `SetupMultiTenantSharding` that sets up the sharding / hybrid multi-tenant feature. This makes it easier to set up this feature.
   - _Easier to change_: You can change the internal services / setting of the sharding / hybrid feature, e.g. one developer wants to store sharding data in a database, instead of in a json file, which this release allows. This is done by creating a new extension method containing the code in the `SetupMultiTenantSharding` extension method with your changes.

## Version 4.1.0

- Adds a service which can handle multiple languages (known in .NET a localization)

_See [Languages & cultures explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Languages-&-cultures-explained) document to see what this new service provides._ 

## Version 4.0.0

- **Now supports .NET 6 and .NET 7 frameworks.**
- All the .NET 6 NuGets have been updated to the latest 6 version.
- Simplification (non-breaking): All tenant version now have a `DataKey` of length of 250.

## Version 3.5.0

Small changes, some of which break existing projects but with very simple rename will fix it. Here is the changes

- BREAKING CHANGE (small): The DisableJwtRefreshToken service has been updated to handle multiple logins from one user. This means that the AuthP refresh token feature can properly handle a single user logging in multiple times.
   - To update your code change the call to `MarkJwtRefreshTokenAsUsedAsync(string userId)` to `LogoutUserViaRefreshTokenAsync(string refreshToken)`. See the `logout` Web API in the [AuthenticateController](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/Example2.WebApiWithToken.IndividualAccounts/Controllers/AuthenticateController.cs).
   - There is also a new method called `LogoutUserViaUserIdAsync(string userId)` that makes all the refresh token as invalid, so all logins for the user with the given userId will timeout.
- BREAKING CHANGE (small): Changed TenantChangeCookieEvent name to SomethingChangedCookieEvent
   - To update your code change the name `TenantChangeCookieEvent` to `SomethingChangedCookieEvent`
- Improved feature: AuthPermissionsDbContext now takes multiple IDatabaseStateChangeEvent
- Improved feature: No AuthP database event change listeners will be triggered during bulk loading

It also updates the [Net.DistributedFileStoreCache](https://github.com/JonPSmith/Net.DistributedFileStoreCache) to version 1.1.0, which is used in AuthP's Example2 (Web API app) to allow the claims in JWT Token to be changed. See article[ASP.NET Core: Three ways to refresh the claims of a logged-in user](https://www.thereformedprogrammer.net/asp-net-core-three-ways-to-refresh-the-claims-of-a-logged-in-user/) for how this works (examples in the version 3.5.0 code).

## Version 3.4.0

This contains one significant new feature and quite a few changes. The most important changes are:

####  New "down for maintenance" feature

This provides a multi-tenant app stop users accessing a tenant while its data is being changed, moved, e.g. moving a tenant's data from a shared database to its own database (sharding). This stops tenant data from being corrupted by a user while the move / change is running. 

See the ["Down for maintenance" feature](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/%22Down-for-maintenance%22-feature) documentation for more information.

#### Added timeout on user invitation

@idan-h said that the Invite new user feature should have a timeout on the invite sent to a person. This was added in this version and the [Invite new user service](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Invite-new-user-service) documentation has been updated to show this.

#### Breaking change to the refresh users claims feature

The refresh users claims feature now uses the [FileStore distributed database](https://www.thereformedprogrammer.net/a-net-distributed-cache-with-a-25-nanosecond-read-time/) for better performance. The the updated [Update claims on tenant change](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Update-claims-on-tenant-change) documentation shows how to add the FileStore distributed database feature.

#### Other changes

There are seven other changes in version 3.4.0 - see the [ReleaseNotes](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/ReleaseNotes.md) for information on these changes.

## Version 3.3.0

The main parts of version 3.3.0 is to create a generic version of the [“invite a user”](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Invite-new-user-service) feature and the [“sign up / versioning”](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sign-up-for-a-new-tenant%2C-with-versioning) features. This also requires two other supporting services:

- A generic way to build an [add new user manager](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Add-New-User-adapter).
- An interface for the “sign up / versioning” service with is used to select the correct database and/or database server for the new tenant - see [this documentation](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Managing-sharding-databases#igetdatabasefornewtenant-service) for more info.

This release also contains some bug fix #35 and improvement #39,

## Version 3.2.0

A study of the version 3.0.0's sharding feature showed that it would very hard to manage multiple databases in production. This version solves this problem by using the `appsettings` file's "ConnectionStrings" section to define the database servers and provides a new `sharingssettings.json` file to manage the databases - see issue #29 for an overview of what is changed. New documentarian are:

- [Multi-tenant config -> Sharding database settings](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sharding-database-settings). 
- [SupportCode -> Managing sharding databases](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Managing-sharding-databases).

Other changes are:

- Improvement: The AuthUserAdmin method called `UpdateUserAsync` now allows you to select which properties you want to update - see docs: Admin -> AuthUser admin.
- Removed: Removed AuthUserAdmin methods 1AddRoleToUser` and `RemoveRoleToUser` as the change to the `UpdateUserAsync` covers this.

## Version 3.1.0

- New feature: Now supports PostgreSQL database
- NOTE: A lot of namespaces changed to support SqlServer and PostgreSQL

## Version 3.0.0

- New (big) Feature: Adding optional sharding to either a single-level or hierarchical multi-tenant applications - see [Defining the three multi-tenant database arrangements](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#2-defining-the-three-multi-tenant-database-arrangements) section in the "multi-tenant explained" docs.
- New (small) Feature: You can mark an AuthUser as disabled, which means no AuthP claims will be added to the user's claims.

BREAKING CHANGES - see [this file](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion3.md) for what you have to change
- The ITenantChangeService has changed to allow multi-tenant sharding to be added - see the UpdateToVersion2.md file for more info
- The option called AppConnectionString has been removed. Its longer needed because of ITenantChangeService change
- Changes to the AuthP database which contains a non-breaking migration. This will be automatically added on startup. 

# Limitations of these release

- Multi-tenant: A user can only have one tenant.

---

# Possible extra features in next release, or future releases

- Add support features
   - Create / delete a database on Azure
   - Add an Azure database to a SQL Server Elastic Pool
   - Stop users from accessing a tenant while its being moved / deleted
   - Update the user’s claims if a tenant is moved / deleted
   - Provide a global “down for maintenance” for tenant users.

---

# OLD VERSIONS

## Version 2.3.1

- New Feature: You can add extra claims to the user via the RegisterAddClaimToUser method
- New Feature: Tap into AuthPermissionsDbContext events by registering a service implementing the IRegisterStateChangeEvent interface
- Fixed potential bug when changing a Roles `RoleType` - see issue #13
- Fixed tenant change problem - see issue #15
- Fixed Hierarchical multi-tenant Move feature problem - see issue #22
- Fixed problem in ClaimCalculator when used with multi-tenant applications - see issue #23

## Version 2.2.0

New Feature: Added "Access the data of other tenant" feature - see issue #10

## Version 2.1.0

This tidies up a few small issues found when writing the "Building ASP.NET Core and EF Core multi-tenant apps"articles. See [ReleaseNotes](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/ReleaseNotes.md) for more info.

## Version 2.0.0

First release of version 2 release of AuthPermissions. This was released early because I found a critical bug fix for hierarchical multi-tenant usage. See [Migrate from AuthPermissions.AspNetCore 1.* to 2.0](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/UpdateToVersion2.md) for detailed information on how to upgrade to this version.

### Summary of new features 

- BREAKING CHANGE: The SetupAspNetCoreAndDatabase configuration method uses a different approach that supports multiple instances of your app
- BREAKING CHANGE: Changes the bulk loading of role and tenants to support the new multi-tenant Roles feature
- BREAKING CHANGE: Updated to net6.0 only
- MULTI-TENANT BREAKING CHANGE: The DataKey format has changed, You need to migrate your application - see issue #12
- MULTI-TENANT BREAKING CHANGE: The RoleAdmin method QueryRoleToPermissions now needs the logged-in userId in multi-tenant applications
- New features: Each multi-tenant can have a different version, e.g. Free, Pro, Enterprise -  see issue #9
- New features: A Tenant Admin user can't see "Advanced Roles", i.e. Role that only an App Admin user should use - see issue #9
- New features: Uses Net.RunMethodsSequentially library to handle startup migrate / seed of databases for applications have multiple instances running

## Version 1.x.x

First release of the AuthPermissions.AspNetCore library, including three example applications.

- [Example1](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example1.RazorPages.IndividualAccounts): an ASP.NET Core Razor Pages application - super-simple introduction to Role-to-Permissions authorization
- [Example2](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example2.WebApiWithToken.IndividualAccounts): an ASP.NET Core Web API application - provides live example of using the JTW refresh token feature.
- [Example3](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example3.MvcWebApp.IndividualAccounts): an ASP.NET Core MVC application - an example of a [single level multi-tenant](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#singlelevel-multi-tenant) application.
- [Example4](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example4.MvcWebApp.IndividualAccounts): an ASP.NET Core MVC application - an example of a [hierarchical multi-tenant](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#hierarchicaltenant-multi-tenant) application.
- [Example5](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/Example5.MvcWebApp.AzureAdB2C): an ASP.NET Core MVC application - an example of using Azure AD for authentication of users.

### Summary of main features

- Provide Role-to-Permissions authorization - see [this article](https://www.thereformedprogrammer.net/finally-a-library-that-improves-role-authorization-in-asp-net-core/) and the [Roles explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Roles-explained) documentation.
- Provide multi-tenant features - see [Multi-tenant explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained).
- Implements an JTW refresh token approach - see [JWT Token refresh explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/JWT-Token-refresh-explained).
- Used NET5. 

### Limitations of this release

- If you have multiple instances of your ASP.NET Core, then you have to manually create / migrate the AuthP's database. See [Version 1 limitations](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Version-1-Limitations) for a fuller explanation. 
- Multi-tenant: A user can only have one tenant.

**Comment by st1pps** [comment]
I just saw this project while thinking about including "better ASP.NET Core authorization" in a new project.
One feature that is missing, and what I'd have to add to that code, are tenant specific roles. The idea is that a tenant admin should be able to create roles for its users, instead of relying on the roles that the system offers. The permissions for these roles would probably be a subset of all available permissions. 
Maybe that would be an interesting addition to this library as well?

**Reply by JonPSmith** [reply]
Hi @st1pps,

I have already implemented an "tenant admin" feature with the following parts:

1. The [List of AuthP's users](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/AuthUser-admin-service#list-authps-users) takes in a DataKey, which will limits the users shown that match that DataKey. So, if you give a tenant users access to the "List Users" code and you pass in the admin user's DataKey it will only shows users in the same tenant as the tenant admin user. It also work well with hierarchical Multi-Tenant as it knows about  the different levels, which means you can have an tenant admin user for "West Coast" and a different admin user for "East Coast" etc. 
   - Have  a look at the [Multi-tenant video at 35 mins in](https://youtu.be/-3hPnGFhY7E?t=2139) which shows this in action.
2. You can mark permissions as ["advanced permissions"](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Setup-Permissions#filtering-out-advanced-permissions), which allows you [filter out advanced permissions](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Roles-admin-service#list-the-permissions-in-your-application), by setting the `excludeFilteredPermissions` to `true` if the admin person has a DataKey. This allows you to control what permissions an tenant admin can add to a role. 
3. Examples 3 and 4 provide an example of a normal admin user, which can access any user, and a tenant admin who can only access users in the same tenant.

PS. I have added a something about [a tenant admin user](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained#the-concept-of-an-tenant-admin-user) in the Concepts -> Multi tenant explained page as its not an obvious feature, which its why you missed it.

**Reply by st1pps** [reply]
Hi Jon,
Thank you for clarifying. This covers a lot of the functionality I need, except for the mentioned tenant specific roles. I'll play around with it a bit, either I'll build something from the "better ASP.NET Core authorization" or I'll fork this project :)

**Reply by JonPSmith** [reply]
Hi @st1pps,

~~Yes, I can see that if a tenant admin created an role it should be limited to their tenant. I will look into this in version 2, which I am working on at the moment (but a lot to do, so it will be some time before its released).~~. See the next comment as to why I won't be adding tenent-level roles.

I have already decided that in version 2 roles that contain "advanced permissions" will not be available to a tenant admin. That removes the the issue of an tenant admin adding a role with powerful features to a tenant user.

One other feature I had in the original "better ASP.NET Core authorization" code was a [PaidForModules](https://github.com/JonPSmith/PermissionAccessControl2/blob/master/PermissionParts/PaidForModules.cs) that allowed you register that certain Permissions are only available to a tenants that have bought that module (see [diagram at the top of the old multi-tenant article](https://www.thereformedprogrammer.net/a-better-way-to-handle-asp-net-core-authorization-six-months-on/)). I haven't implemented that in the AuthP library as a) not many people use that, and b) its a bit clunky as you have to edit the Permissions code. If that's useful then I could add it to AuthP, but I would want a "contribution" (either you write the code and PR it,  or I write the code for money😄) for that work .

Finally a warning from my experience of designing a really big multi-tenant. My client has users that worked for different retail outlets, which meant that when a user that works in multiple retail outlets logged I had to provide a list of the retail outlets they were registered with so that when log in to the correct one. It was a pain to implement and I'm not planning to add it to the AuthP library.

**Reply by JonPSmith** [reply]
Hi @st1pps,

I was thinking about your request and I'm not sure that having a tenant have their own roles. A SaaS (Multi-tenant) application is about providing the same service to many organizations. This suggests that there should be a global set of roles, which the tenant admin can add/remove to tenant users. I also think having roles linked to a tenant would make it very difficult for an overall admin user to work out any problems.

The EF Core team are always talking about not adding features to the EF Core that could cause the developers to get into a bad situation. For this reason I won't be implementing a tenant-only role, but I think the PaidForModules is still relevant, as they are global way to manage the features available to users.

**Reply by st1pps** [reply]
Hi @JonPSmith ,
Thank you for the explanation, I understand the reasoning. I'm in a situation where the tenant admin needs to be able to allow or disallow editing of properties of the tenant's users' profiles. Which properties are allowed to be edited should be configurable by "editing roles", so that the same permissions set can be applied to multiple users. Following your reasoning I'll probably be better off managing these "permissions" separately from the whole authorization system in my domain logic. 
Thank you for answering :)

**Reply by JonPSmith** [reply]
You have to understand that the library must provide a secure system for managing users, so I ensure that there are strong limits on what the tenant admin can do. That doesn't mean you can't write your own code, most likely using the various [Admin services](https://github.com/JonPSmith/AuthPermissions.AspNetCore/tree/main/AuthPermissions/AdminCode/Services), to implement extra features for the tenant admin user. Its then up to you to make sure one tenant user can't change something they shouldn't have access to.

You  _might_ find some of the extra features in version 2.0.0 of this library could help you. For instance I have implemented what I call 'tenant Roles' that can be added to a tenant such that each tenant can have optional versions. One use would have different versions of the app, say Free, Pro, Enterprise, and each tenant can pay for the higher levels. See issue #9 which describes this.

NOTE: I released version 2.0.0 early because I found a critical bug in hierarchical multi-tenant in version 1. I will adding documentation and examples on the new multi-tenant features in the coming weeks.

**Reply by JonPSmith** [reply]
Hi @st1pps,

I have released an article called [Multi-tenant apps with different versions can increase your profits](https://www.thereformedprogrammer.net/multi-tenant-apps-with-different-versions-can-increase-your-profits/) that explains how the AuthP version 2.0.0 allows you to create Roles linked to a tenant. I think that is what you were asking for.

**Comment by mabouchalkha** [comment]
Hi,
Thanks for your time and effort to do this library,
Can we have a support for spa application support like a blazor wasm?

Thank you

**Reply by JonPSmith** [reply]
In theory, yes, but I haven't tried it. A blazor webassembly still has a `Program` class where you register the AuthP services, and once that is done you can access the various services Roles, Tenant and AuthUser service.

BUT the blazor webassembly [doesn't support HostedSevices](https://github.com/dotnet/aspnetcore/issues/7213), which means you can't use AuthP's `SetupAspNetCoreAndDatabase` extension method that migrates the AuthP database on startup. Instead you need to use `SetupAspNetCorePart` extension method an manually migrate the AuthP database. You also can't use the `AddSuperUserToIndividualAccounts` method either, as that uses HostedSevices too and any bulk loading won't work (but bulk loading is only used my me in the examples).

**Comment by fenterbug** [comment]
Have you put any thought into supporting users with multiple roles? I've read the article series but haven't looked at the example app, so I don't know if it's already there. (Sorry)

Use case: Imagine an inventory system targeted to small businesses. You'd probably have a role specific to filling orders (and, thus, decrementing your warehouse inventory) and a different role specific to restocking (incrementing) your warehouse inventory. If you only have a handful of employees, then the person doing the restocking will also probably be filling orders.

Conceptually, people often wear multiple hats... have multiple roles. (I'm both a "developer" as well as "technical support".) I could have a single role that includes the permissions for both "hats", but I feel it's a better interpretation of the analogy to assign multiple "hats", multiple roles, to a single person.

**Reply by JonPSmith** [reply]
Users can have multiple Roles. You should read the [first article](https://www.thereformedprogrammer.net/finally-a-library-that-improves-role-authorization-in-asp-net-core/) and look at the diagram at the top. You will see that the user has two Roles: Sales and FirstAid.

**Comment by klasyc** [comment]
Hello, I like the AuthP library, but I miss the support of PostreSQL. Is it possible or planned extra feature to the next release?

**Reply by JonPSmith** [reply]
Hi @klasyc,

PostgreSQL is in the roadmap, but a number of people have asked for sharding for multi-tenant applications. I plan to implement sharding first, then I will look at PostgreSQL.

**Reply by JonPSmith** [reply]
Version 3.1.0 supports PostgreSQL

**Comment by akema-trebla** [comment]
Hi @JonPSmith 

I love this library very much. It's got a lot of things I've wanted in a multi-tenant application/ library. 

I would like to contribute to it, especially with enabling users to manage multiple the tenants. 

It seems you had some ideas already. How do you propose we communicate and so I can help?

**Reply by JonPSmith** [reply]
Hi @akema-trebla,

I'm pleased you like this library and humbled that you would like contribute this library. I think the feature you suggested was allowing a  user to be linked to multiple tenants. I wouldn't suggest that because I have already implemented that for the original client. The main problem is making a generic feature that will would with any authentication handler, which is a bit complex.

Have you build a multi-tenant app yet? Maybe you would like to improve something in the  Example3 or Example4 multi-tenant applications (I'm improving Example6 to add some new support features).

**Reply by JonPSmith** [reply]
@akema-trebla, come back to me with any suggestions before you actually change things as there are a few things to think about, e.g. adding more tests

**Reply by akema-trebla** [reply]
Hi @JonPSmith,

Sorry for the late response. I've been unavailable for a while but I'm back now. 

I built a multi-tenant fashion designers' mobile app a few years ago and I decided to make it more scalable and easier to manage and that's when I found this project. I think it's a great productivity boost to have and I'd be honored to help in anyway I can.

Does AuthP have a separate nuget package for the BaseCode? I'm asking because of DDD considerations.
Also is there any work needed to use sharding with with a hierarchical tenant or can we go ahead with an example there?
And anything else you would like me to work on.

You can shoot me an email at albertameka@gmail.com.

**Reply by JonPSmith** [reply]
Hi @akema-trebla,

> Does AuthP have a separate nuget package for the BaseCode?

The AuthP code is contained in multiple projects which all start with `AuthPermissions.` and I use a dotnet tool called `JonPSmith.MultiProjPack` to create a composite NuGet. I describe how to do that in [this section](https://github.com/JonPSmith/AuthPermissions.AspNetCore#notes-on-creating-a-nuget-package) of the README.md file. But for adding examples you don't need that.

> Also is there any work needed to use sharding with with a hierarchical tenant or can we go ahead with an example there?

You only need to set the AuthP's options with `options.TenantType = TenantTypes.HierarchicalTenant | TenantTypes.AddSharding;`.

> And anything else you would like me to work on.

I don't think an extra example of hierarchical tenant + sharding is worth the effort, as its just a mix of Example4 (hierarchical) and Example6 (Sharding). 

What are your skills - do they cover the front-end as well as a backend? An example of using [Blazor Server](https://docs.microsoft.com/en-us/aspnet/core/blazor/hosting-models#blazor-server) with AuthP would be interesting. You could take Example4 (hierarchical) and create a new example that uses Blazor Server. That's a lot of work but a lot of people are asking how you would use AuthP with Blazor.

If that is too much then I understand.

**Reply by akema-trebla** [reply]
Alright. I'll work on the Blazor example.

**Reply by JonPSmith** [reply]
Hi @akema-trebla,

That's great! Its quite a bit of work and a few things, like authorization in Blazor, will need some help from me.

Can I suggest you open a new issue for this topic so we can confer on anything. I use issues when I find something I want to do and I might put down my thoughts (see issue #29) or to list the things I need to do (see issue #35).

**Comment by nsfuntionality** [comment]
Hello Jon, 
Is there a future plan to introduce something like RabbitMQ for fewer calls to DB for any old data that will not change? this way instead of a DB call, something like RabbitMQ or something can respond.
Any Plan on the Feature for data retention? so the DB is just not growing crazy with old data?

**Reply by JonPSmith** [reply]
Hi @nsfuntionality,

The AuthP library provides a structure for building apps with administration features, but in normal use the AuthP code doesn't access the database because it uses claims. 

If you want to improve the performance of your application, then you need look at the code from your part of the application, e.g. Example3 provides a Invoice handler and that is the part that you need to apply caching, RabbitMQ or whatever to the part of your code. 

You might like to read my [Five levels of performance tuning for an EF Core query](https://www.thereformedprogrammer.net/five-levels-of-performance-tuning-for-an-ef-core-query/) article or chapters 14 to 16 in my book [Entity Framework Core in Action](https://bit.ly/EfCoreBookEd2) for various ways to improve database-heavy applications.

**Comment by ccnicholls99** [comment]
I like this approach very much indeed. Did my own implementation for ongoing app development using ASP identity but your ideas are much better formed than mine so I'm switching to your plan.

Some thoughts on Tenant Creation...
- 'versions' could be a little more developed. Maybe a formal interface added to AuthPermissions.SupportCode.AddUsersServices.AddNewTenantDto, version parameter, instead of just a string?
- Basic properties of key/name, description, features for display in the UI (Controllers.HomeController.CreateTenant) rather than hard-coded content.
- option for model-specific 'limits' per version based on usage....
- i.e. Version has max n-Tenants, max n-Users, 
- i.e. Tenant Data has max n-Employees, max n-Departments, etc (or whatever makes sense in the business domain of the tenant). 


Hope that makes sense.
Keep up the good work!

**Comment by nsfuntionality** [comment]
Is there way we can have sub domain once the tenant is logged in? just like how we have navbar brand change to tenant name based on the login?

**Reply by JonPSmith** [reply]
Hi @nsfuntionality,

> Is there way we can have sub domain once the tenant is logged in?

I think you want the tenant name. The most straight way is to get the user's `UserId` from the claims and look up the `AuthUser` using the `IAuthUsersAdminService.FindAuthUserByUserIdAsync(userId)`. This returns the `AuthUser` with its `UserTenant`, which has a `TenantFullName` properly holding the tenant's name. The problem with this approach is if you call this every HTTP request it will slow down the performance of your app.

There is an AuthP feature called `RegisterAddClaimToUser<>` that's adds a new claim on login - see [this section](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Startup-code#1-registeraddclaimtouser). This is much quicker as the claims are available. _NOTE: The new 6.2.0 "Sign up for a new tenant, with versioning" feature uses a interim tenant name, so you need to add extra code to update the claims - see [Update claims on tenant change](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Update-claims-on-tenant-change)._

Another useful article is [ASP.NET Core: Three(+1) ways to refresh the claims of a logged-in user](https://www.thereformedprogrammer.net/asp-net-core-three-ways-to-refresh-the-claims-of-a-logged-in-user/), which gives some different approaches.
